### PR TITLE
feat: add kafka migration

### PR DESCRIPTION
**Pull request by johnjcasey**
Adds the tooling to support migrating a PubSub lite topic to a Managed Kafka topic

### CODE REVIEW

File: migration/docker/Dockerfile [hunk 1]
```diff
@@ -0,0 +1,85 @@
+FROM --platform=linux/amd64 eclipse-temurin:21
+ARG KAFKA_VERSION="3.4.0"
+ARG KAFKA_CONNECT_VERSION="${KAFKA_VERSION}"
+ARG KAFKA_SCALA_VERSION="2.13"
+ARG PUBSUB_GROUP_KAFKA_CONNECTOR_VERSION="1.2.0"
+ARG KAFKA_HOME_ROOT="/opt"
+ENV KAFKA_HOME="${KAFKA_HOME_ROOT}/kafka"
+ARG KAFKA_CONFIG_DIR="${KAFKA_HOME}/config"
+ARG KAFKA_RELEASE="kafka_${KAFKA_SCALA_VERSION}-${KAFKA_VERSION}"
```
Comment (rajanya-google) [inline migration/docker/Dockerfile:9]: why is KAFKA_SCALA_VERSION included in the KAFKA_RELEASE?

File: migration/docker/Dockerfile [hunk 2]
```diff
@@ -0,0 +1,85 @@
...
+# The pubsub-group-kafka-connector file needs to be pre-built/downloaded using maven or other similar tool.
+# References:
+# 1) https://github.com/googleapis/java-pubsub-group-kafka-connector/releases/
+# 2) https://central.sonatype.com/artifact/com.google.cloud/pubsub-group-kafka-connector
+ARG PUBSUB_GROUP_KAFKA_CONNECTOR_JAR="pubsub-group-kafka-connector-${PUBSUB_GROUP_KAFKA_CONNECTOR_VERSION}.jar"
+ARG KAFKA_CONNECT_CONFIGURE_SCRIPT="configure-kafka-connect.sh"
+ARG BUILD_KAFKA_CONNECT_STARTUP_SCRIPT="start-kafka-connect.sh"
+ARG BUILD_PSL_JOB_STARTUP_SCRIPT="start-psl-connector.sh"
+ARG BUILD_KAFKA_CONNECT_CONFIG_FILE="kafka-connect.properties"
+ARG BUILD_PUBSUB_LITE_JOB_FILE="PSL_job.json"
```
Comment (rajanya-google) [inline migration/docker/Dockerfile:21]: rename to psl_job.json ?

File: migration/docker/Dockerfile [hunk 3]
```diff
@@ -0,0 +1,85 @@
...
+ENV KAFKA_HEAP_OPTS="-Xms2G -Xmx2G"
+ENV KAFKA_HOME="${KAFKA_HOME_ROOT}/kafka"
```
Comment (rajanya-google) [inline migration/docker/Dockerfile:23]: it was declared in line#7

File: migration/docker/Dockerfile [hunk 4]
```diff
@@ -0,0 +1,85 @@
...
+ARG KAFKA_PLUGINS_DIR="${KAFKA_HOME}/plugins"
```
Comment (rajanya-google) [inline migration/docker/Dockerfile:24]: Please move KAFKA_ related ARG together.

File: migration/docker/Dockerfile [hunk 5]
```diff
@@ -0,0 +1,85 @@
...
+ARG KAFKA_RELEASE="kafka_${KAFKA_SCALA_VERSION}-${KAFKA_VERSION}"
+ARG KAFKA_TARBALL="${KAFKA_RELEASE}.tgz"
+ARG KAFKA_DOWNLOAD_URL="https://archive.apache.org/dist/kafka/${KAFKA_VERSION}/${KAFKA_TARBALL}"
+# The pubsub-group-kafka-connector file needs to be pre-built/downloaded using maven or other similar tool.
+# References:
+# 1) https://github.com/googleapis/java-pubsub-group-kafka-connector/releases/
+# 2) https://central.sonatype.com/artifact/com.google.cloud/pubsub-group-kafka-connector
+ARG PUBSUB_GROUP_KAFKA_CONNECTOR_JAR="pubsub-group-kafka-connector-${PUBSUB_GROUP_KAFKA_CONNECTOR_VERSION}.jar"
+ARG KAFKA_CONNECT_CONFIGURE_SCRIPT="configure-kafka-connect.sh"
+ARG BUILD_KAFKA_CONNECT_STARTUP_SCRIPT="start-kafka-connect.sh"
```
Comment (rajanya-google) [inline migration/docker/Dockerfile:18]: why BUILD_ prefix?

File: migration/docker/Dockerfile [hunk 6]
```diff
@@ -0,0 +1,85 @@
...
+    --mount=type=secret,id=kafka_ssl_truststore_location \
+    --mount=type=secret,id=kafka_ssl_truststore_password \
+    KAFKA_CONNECT_CONFIG_FILE="${KAFKA_CONNECT_CONFIG_FILE}" \
+    KAFKA_BOOTSTRAP_SERVERS="$(cat /run/secrets/gmk_bootstrap_servers)" \
+    KAFKA_SASL_SERVICE_ACCOUNT="$(cat /run/secrets/gmk_sasl_service_account)"\
+    KAFKA_SASL_SERVICE_ACCOUNT_KEY="$(cat /run/secrets/gmk_sasl_service_account_key)" \
+    KAFKA_SINK_TOPIC="$(cat /run/secrets/kafka_sink_topic)" \
+    KAFKA_CONNECT_GROUP_ID="$(cat /run/secrets/kafka_connect_group_id)" \
+    KAFKA_PLUGINS_DIR=${KAFKA_PLUGINS_DIR} \
+    PUBSUB_LITE_GCP_PROJECT="$(cat /run/secrets/pubsub_lite_gcp_project)" \
```
Comment (rajanya-google) [inline migration/docker/Dockerfile:72]: let us replace PUBSUB_LITE with PSL or vice versa at all places.

File: migration/k8s.yaml [hunk 7]
```diff
@@ -0,0 +1,44 @@
...
+    matchLabels:
+      app: "<workflow_name>"
+  template:
+    metadata:
+      labels:
+        app: "<workflow_name>"
+    spec:
+      serviceAccountName: <gke_service_account>
+      containers:
+      - name: "psl-to-gmk-1"
```
Comment (rajanya-google) [inline migration/k8s.yaml:21]: why psl-to-gmk-1?

File: migration/docker/Dockerfile (hunk 1 above)
Comment (johnjcasey) [reply migration/docker/Dockerfile:9]: Most of these came pre-configured with the work Pramod did. I'm not entirely sure here

File: migration/docker/Dockerfile (hunk 5 above)
Comment (johnjcasey) [reply migration/docker/Dockerfile:18]: There is the location relative to the script, and the location within the docker image. BUILD is the first

File: migration/k8s.yaml (hunk 7 above)
Comment (johnjcasey) [reply migration/k8s.yaml:21]: The container needs a name, this is the one that the yaml populated by default

**Review by samarthsingal — COMMENTED** [review]
I think the changes broadly look good but I'm wondering if the migration/ folder itself needs to live in this project. This repo is mostly used for self-managed deployments.

**Comment by rajanya-google** [comment]
LGTM

**Review by rajanya-google — APPROVED** [review]

**Pull request merged**